### PR TITLE
IGNITE-22163 Fix potential NPE in IndexManager

### PR DESCRIPTION
**Pull request by alievmirza**
https://issues.apache.org/jira/browse/IGNITE-22163

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java [hunk 1]
```diff
@@ -180,14 +180,19 @@ private CompletableFuture<?> onPrimaryReplicaElected(PrimaryReplicaEventParamete
...
-                return getMvTableStorageFuture(parameters.causalityToken(), primaryReplicaId)
-                        .thenCompose(mvTableStorage -> awaitPrimaryReplica(primaryReplicaId, parameters.startTime())
-                                .thenAccept(replicaMeta -> tryScheduleBuildIndexesForNewPrimaryReplica(
+                CompletableFuture<ReplicaMeta> replicaMetaFuture = awaitPrimaryReplica(primaryReplicaId, parameters.startTime());
+
+                CompletableFuture<MvTableStorage> mvTableStorageFuture =
+                        getMvTableStorageFuture(parameters.causalityToken(), primaryReplicaId);
+
+                return replicaMetaFuture
+                        .thenAcceptBoth(mvTableStorageFuture, (replicaMeta, mvTableStorage) ->
```
Comment (tkalkirill) [inline modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java:189]: Why do you think that we need to wait for these two futures in parallel?
Let's say `mvTableStorageFuture` returns `null`, then it will be NPE, how do you think this will fix the problem from the ticket?

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManager.java [hunk 2]
```diff
@@ -171,8 +171,8 @@ public CompletableFuture<Void> stopAsync() {
      * @return Future with multi-version table storage, completes with {@code null} if the table does not exist according to the passed
      *      parameters.
      */
-    CompletableFuture<MvTableStorage> getMvTableStorage(long causalityToken, int tableId) {
-        return tableManager.tableAsync(causalityToken, tableId).thenApply(table -> table.internalTable().storage());
+    CompletableFuture<@Nullable MvTableStorage> getMvTableStorage(long causalityToken, int tableId) {
```
Comment (tkalkirill) [inline modules/index/src/main/java/org/apache/ignite/internal/index/IndexManager.java:174]: Missing junit test for `null` case.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java [hunk 3]
```diff
@@ -250,8 +277,8 @@ private void stopBuildingIndexesIfPrimaryExpired(TablePartitionId replicaId) {
         }
     }
 
-    private CompletableFuture<MvTableStorage> getMvTableStorageFuture(long causalityToken, TablePartitionId replicaId) {
-        return indexManager.getMvTableStorage(causalityToken, replicaId.tableId());
+    private CompletableFuture<MvTableStorage> getMvTableStorageFuture(long causalityToken, int tableId) {
```
Comment (tkalkirill) [inline modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java:280]: Why did you change it to `int tableId`, before that it wasn’t bad.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java [hunk 4]
```diff
@@ -143,14 +145,27 @@ private CompletableFuture<?> onIndexBuilding(StartBuildingIndexEventParameters p
...
-                                            indexDescriptor,
-                                            mvTableStorage,
-                                            replicaMeta
-                                    ))
+                    int tableId = indexDescriptor.tableId();
+
+                    CompletableFuture<?> startBuildIndexFuture = getMvTableStorageFuture(parameters.causalityToken(), tableId)
+                            .thenCompose(mvTableStorage -> {
+                                if (mvTableStorage == null) {
+                                    throw new IgniteInternalException(
```
Comment (tkalkirill) [inline modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java:153]: I think it will be enough for us to add a `requireMvTableStorageNonNull `and throw an exception with the general text; the stack trace will make it clear where the error occurred.

You could also keep the existing code and add something like:
```
    private CompletableFuture<MvTableStorage> getMvTableStorageFuture(long causalityToken, TablePartitionId replicaId) {
        return indexManager.getMvTableStorage(causalityToken, replicaId.tableId())
                .thenApply(IndexBuildController::requireMvTableStorageNonNull);
    }

    private static MvTableStorage requireMvTableStorageNonNull(MvTableStorage mvTableStorage) {
        ...
    }
```

Comment (alievmirza) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java]: good proposal, thanks!

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java (hunk 3 above)
Comment (alievmirza) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java:258]: because we need only tableId for `org.apache.ignite.internal.index.IndexManager#getMvTableStorage`, and semantically we use table id, not replica id, to get MvTableStorage. I don't see a reason to pass `TablePartitionId replicaId` to `getMvTableStorageFuture` then

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManager.java (hunk 2 above)
Comment (alievmirza) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManager.java:175]: added

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java (hunk 1 above)
Comment (alievmirza) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java]: reverted, you are right

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuildControllerTest.java [hunk 5]
```diff
@@ -167,6 +172,22 @@ void testStartBuildIndexesOnIndexBuilding() {
         );
     }
 
+    @Test
+    void testExceptionIsThrownOnIndexBuildingWhenStorageIsNull() {
+        setPrimaryReplicaWhichExpiresInOneSecond(PARTITION_ID, NODE_NAME, NODE_ID, clock.now());
+
+        createIndex(INDEX_NAME);
+
+        when(indexManager.getMvTableStorage(anyLong(), anyInt())).thenReturn(completedFuture(null));
```
Comment (tkalkirill) [inline modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuildControllerTest.java:181]: Please use `CompletableFutures#nullCompletedFuture`

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuildControllerTest.java [hunk 6]
```diff
@@ -187,6 +208,20 @@ void testStartBuildIndexesOnPrimaryReplicaElected() {
         );
     }
 
+    @Test
+    void testExceptionIsThrownOnPrimaryReplicaElectedWhenStorageIsNull() {
+        when(indexManager.getMvTableStorage(anyLong(), anyInt())).thenReturn(completedFuture(null));
```
Comment (tkalkirill) [inline modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuildControllerTest.java:213]: Please use `CompletableFutures#nullCompletedFuture`

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuildControllerTest.java (hunk 5 above)
Comment (alievmirza) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuildControllerTest.java]: done

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuildControllerTest.java (hunk 6 above)
Comment (alievmirza) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuildControllerTest.java]: done

**Review by tkalkirill — APPROVED** [review]
